File: 2018/src/q1.cpp
#include <fstream>
#include <iostream>
#include <vector>
#include <string>
#include <unordered_set>

int addNums(std::vector<int> data) {
  int res = 0;
  for(int num : data) res += num;
  return res;
}

int duplicateFrequency(std::vector<int> data) {
  std::unordered_set<int> freqs = {0};
  int currFreq = 0;
  size_t index = 0;
  bool found = false;

  while(!found) {
    currFreq += data[index++];
    if(index == data.size()) index = 0;

    auto search = freqs.find(currFreq);
    if(search != freqs.end()) {
      found = true;
    } else {
      freqs.insert(currFreq);
    }
  }

  return currFreq;
}

int main() {
  std::ifstream ifstrm("./data/q1.txt", std::ios::in);
  std::vector<int> data;

  if(!ifstrm.is_open()) {
    std::cout << "Failed to open file" << std::endl;
  } else {
    std::string line;
    while(std::getline(ifstrm, line)) {
      line[0] == '+' ? data.push_back(stoi(line.substr(1))) : 
        data.push_back(stoi(line.substr(1)) * -1);
    }
  }

  std::cout << "Answer to part 1: " << addNums(data) << std::endl;
  std::cout << "Answer to part 2: " << duplicateFrequency(data) << std::endl;
  return 0;
}

File: 2020/q1.py
TARGET = 2020

def part1(nums):
    currNums = nums[:]
    currNums.sort()

    i = 0
    j = 0
    length = len(currNums)
    found = False

    while i < length and not found:
        while j < length and (currNums[i] + currNums[j] < TARGET):
            j += 1
        if (currNums[i] + currNums[j]) == TARGET:
            found = True
        else:
            i += 1
            j = i
    return currNums[i] * currNums[j]

def part2(nums):
    currNums = nums[:]
    currNums.sort()

    i = 0
    j = 1
    k = 2
    length = len(currNums)
    found = False

    while i < length - 2 and not found:
        while j < length - 1 and not found:
            while k < length and (currNums[i] + currNums[j] + currNums[k] < TARGET):
                k += 1
            if (currNums[i] + currNums[j] + currNums[k]) == TARGET:
                found = True
            else:
                j += 1
                k = j + 1
        if not found:
            i += 1
            j = i + 1
            k = j + 1
    
    return currNums[i] * currNums[j] * currNums[k]
nums = []

with open("./data/q1.txt", "r") as file:
    for line in file:
        nums.append(int(line.strip()))

print("Answer to part1: {}".format(part1(nums)))
print("Answer to part2: {}".format(part2(nums)))